src/pTSL.py

```diff
@@ -1,36 +1,263 @@
+
+        return sse
+
+    def train(self, corpus_file):
+        # [[['a','b'],1],[]]
+        corpus_probs = self.read_corpus_file(corpus_file, True)
+        # get list of tiers
+        # from initialized ptsl (self.probs and self.factor) and corpus data, create complete list of alphabet (no duplicates)
+        param = {x for sublist in corpus_probs for x in sublist[0]}
+        # remove fixed alphabets
+        param = list(param - {x for x in self.probs})
+
+        # create bounds
+        # instead of limiting bound for fixed value, I removed it completely from the parameter
+        bounds = [(0, 1) for i in range(len(param))]
+
+        # bounds = []
+        # for t in param:
+        #     if t in self.probs:
+        #         bounds.append((self.probs[t], self.probs[t]))
+        #     else:
+        #         bounds.append((0, 1))
+
+        # randomly initialize parameter - this will be the input
+        random_proj = rand(len(param))
+        # run the minimize function
+        proj_res = minimize(self.evaluate_proj,
+                            random_proj,
+                            bounds=bounds,
+                            method='L-BFGS-B',
+                            args=(param, corpus_probs))
+
+        optimal_projection = dict(zip(param, proj_res.x))
+        self.probs.update(optimal_projection)
+
+    def print_projections(self):
+        for key in sorted(self.probs.keys()):
+            print("{}: {}".format(key, round(self.probs[key], 4)))
+
+    ''' 
+    initialize pTSL based on the given path to files (tier, prob, grammar)
+    '''
+    @classmethod
+    def from_file(cls, tier_file, factors_file, verbose=False):
+        '''
+            An alternative constructor that creates a pTSL object based on
+            the contents of input files. The files should have the following formats:
+
+            TIER_FILE
+            <symbol 1> <projection 1>
+            <symbol 2> <projection 2>
+            ...
+            <symbol n> <projection n>
+
+            FACTORS_FILE
+            <factor 1>
+            <factor 2>
+            ...
+            <factor n>
+
+
+            <factor i> is a segment separated by spaces.
+            Segments do not need to be single characters.
+
+            Input:
+                tier_file: A string specifying where to find the tier file.
+                factors_file: a string specifying where to find the factor file
+        '''
+
+        probs = {}
+        with open(tier_file, 'r') as csv_file:
+            csv_reader = csv.reader(csv_file)
+            for row in csv_reader:
+                if len(row) == 1:
+                    probs[row[0]] = 0
+                else:
+                    probs[row[0]] = float(row[1])
+
+        factors = []
+        with open(factors_file, 'r') as csv_file:
+            csv_reader = csv.reader(csv_file)
+            for row in csv_reader:
+                factors.append(tuple(row[0].split()))

```

Comment:
Might be useful to add a warning if the length of every factor is 1, which probably indicates that the user forgot to put in spaces.